java/ql/lib/semmle/code/java/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -3,6 +3,25 @@ private import DataFlowImplSpecific::Public
 import Cached
 
 module DataFlowImplCommonPublic {
+  /** A state value to track during data flow. */
+  abstract class FlowState extends string {
```

Comment:
This is a potential foot gun: For data flow configurations in libraries like https://github.com/github/codeql/blob/c5728b2951560649b158b330e5051d8fe1215695/cpp/ql/lib/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll#L737
if someone were to add new states to a query (which I assume would be a valid thing to do), then such library analyses would have to be recomputed.

Can't `FlowState` just be an alias for `string`?

Reply:
Good point. There might be one place in DataFlowImpl where this change will require a refactor.

Reply:
Maybe this has an obvious answer, but isn't codeql-javascript having the same potential foot gun?
Or is the plan to (later) integrate this PR with the javascript data-flow library?
https://github.com/github/codeql/blob/15caaa7ad6d91c533596204b6a863a1c56240628/javascript/ql/lib/semmle/javascript/dataflow/Configuration.qll#L284-L293

Reply:
Better to let @github/codeql-javascript answer that.

Reply:
I don't think the javascript data-flow library supports usage in library code - it only exists in one copy, so that should be reserved for queries. It's a good observation and a good question, though, and certainly something that needs to be taken into account if the javascript data-flow library is to allow in-library use.

Reply:
> I don't think the javascript data-flow library supports usage in library code - it only exists in one copy, so that should be reserved for queries. 

Yep. We only create new `FlowLabel`s in query specific code, and never in library code.   
We also have [a test that fails](https://github.com/github/codeql/blob/main/javascript/ql/test/library-tests/FlowLabels/DefaultFlowLabels.ql) if we every accidentally add more default flow labels.

Reply:
The problem is not whether you create flow labels in library code, but whether you use global data flow in library code.

Reply:
Ahh. We don't use global dataflow in library code, we use type-tracking instead.

Reply:
I'm supportive of the idea of using an alias `class FlowState = string` instead of the abstract class.

In practice you want to define a predicate for accessing the flow state constant. Once you have that predicate, you don't gain much by forcing users to declare a subclass as well. In JS this has led to various foot-gun situations with [awkward solutions](https://github.com/github/codeql/blob/main/javascript/ql/lib/semmle/javascript/security/dataflow/TaintedPathQuery.qll#L14).